### PR TITLE
improve RGB5x52Gray()

### PR DESCRIPTION
**Pull request by K-Shinotsuka**
improve RGB5x52Gray()

### CODE REVIEW

File: modules/imgproc/src/color.cpp [hunk 1]
```diff
@@ -1141,37 +1135,35 @@ struct RGB5x52Gray
             #elif CV_SSE2
             if (haveSIMD)
             {
-                __m128i v_zero = _mm_setzero_si128();
-
+                const __m128i v_bg2y = _mm_set1_epi32((G2Y << 16) + B2Y);
+                const __m128i v_r2yd = _mm_set1_epi32((1 << 16) + R2Y);
+                const short delta = 1 << (yuv_shift - 1);
+                const __m128i v_delta = _mm_set1_epi32((delta << 16) + delta);
```
Comment (ilya-lavrenov) [inline modules/imgproc/src/color.cpp]: can you use `_mm_set1_epi16` here and above?

File: modules/imgproc/src/color.cpp [hunk 2]
```diff
@@ -1205,37 +1197,35 @@ struct RGB5x52Gray
...
```
Comment (ilya-lavrenov) [inline modules/imgproc/src/color.cpp]: the same.

**Comment by K-Shinotsuka** [comment]
@ilya-lavrenov 
Sorry. I have not finished improving the code.
I will finish improving the code and measuring the speed in 2016/7/23 or 2016/7/24.
Please wait a moment.

**Comment by K-Shinotsuka** [comment]
@ilya-lavrenov 

Though it's earlier than planned, I have finished improving the code and measuring the clocks.
Please observe the following.

<table border=1>
 <tr><th>green</th><th>width</th><th>Average Clocks (original code)</th><th>Average Clocks (new code)</th><th>Faster Rate (compared with original code)</th></tr>
 <tr><td>5</td><td>8</td><td>135.0</td><td>102.8</td><td>31.3%</td></tr>
 <tr><td>5</td><td>16</td><td>197.4</td><td>138.6</td><td>42.4%</td></tr>
 <tr><td>5</td><td>32</td><td>239.2</td><td>203.2</td><td>17.7%</td></tr>
 <tr><td>5</td><td>64</td><td>370.4</td><td>273.6</td><td>35.4%</td></tr>
 <tr><td>5</td><td>128</td><td>676.6</td><td>515.0</td><td>31.4%</td></tr>
 <tr><td>6</td><td>8</td><td>172.8</td><td>117.8</td><td>46.7%</td></tr>
 <tr><td>6</td><td>16</td><td>176.6</td><td>161.2</td><td>9.6%</td></tr>
 <tr><td>6</td><td>32</td><td>251.0</td><td>226.0</td><td>11.1%</td></tr>
 <tr><td>6</td><td>64</td><td>383.8</td><td>277.2</td><td>38.5%</td></tr>
 <tr><td>6</td><td>128</td><td>674.6</td><td>488.4</td><td>38.1%</td></tr>
</table>

Measurement Conditions
OS : Ubuntu 16.04
Compiler : g++ (Ubuntu 5.3.1-13ubuntu3) 5.3.1 20160330
CPU : Intel(R) Core(TM)2 Duo CPU E8500 @ 3.16GHz
Compiler Option : -msse2 -O3
Measurement Function: rdtsc()

File: modules/imgproc/src/color.cpp [hunk 3]
```diff
@@ -1205,37 +1199,30 @@ struct RGB5x52Gray
...
-                                           _mm_unpackhi_epi16(v_mullo_r, v_mulhi_r));
-
-                    v_dst0 = _mm_srli_epi32(v_dst0, yuv_shift);
-                    v_dst1 = _mm_srli_epi32(v_dst1, yuv_shift);
-
-                    __m128i v_dst = _mm_packs_epi32(v_dst0, v_dst1);
-                    _mm_storel_epi64((__m128i *)(dst + i), _mm_packus_epi16(v_dst, v_zero));
+                    __m128i v_b = _mm_srli_epi16(_mm_slli_epi16(v_src, 11), 8),
+                            v_g = _mm_srli_epi16(_mm_slli_epi16(_mm_srli_epi16(v_src, 5), 11),8),
+                            v_r = _mm_srli_epi16(_mm_slli_epi16(_mm_srli_epi16(v_src, 10), 11),8);
```
Comment (vpisarev) [inline modules/imgproc/src/color.cpp]: is it really necessary to use 3 shifts? I thought, 2 shifts could be enough

**Comment by K-Shinotsuka** [comment]
@vpisarev 

>is it really necessary to use 3 shifts? I thought, 2 shifts could be enough

Is it "v_r"?
b is 0-4bit.
g is 5-9bit.
r is 10-14bit.
Is 15bit's value always 0?

**Comment by vpisarev** [comment]
@K-Shinotsuka, sorry for delay, really lot's of meetings :)
I mean, you have the code, say, for v_g (v_b and v_r are similar):

<pre>
((v_src >> 5) << 11) >> 8 
</pre>

I'm pretty sure it can be done with 2 shifts, e.g. `(v_src >> 5) << 3`

**Comment by K-Shinotsuka** [comment]
@vpisarev 

I did as you said. I changed v_g.
Error is occurred at Size_CvtMode_cvtColor8u.cvtColor8u.

Again, 
b is 0-4bit.
g is 5-9bit.
r is 10-14bit.
15bit's value is not always 0.
So, if the number of shifts is 2, the calculation is wrong.

**Comment by terfendail** [comment]
Could we use `(v_src & 0x3E0) >> 2` to extract v_g ?

**Comment by vpisarev** [comment]
@K-Shinotsuka, oh, right, my fault, sorry. So you can revert to your original code or try @terfendail's suggestion.

**Comment by K-Shinotsuka** [comment]
@vpisarev 

I understand. I revert to my original code.

@terfendail 

Thank you for your advice.
But I have already tried it and that's code is slower than the code that I have committed.
So, I revert to my original code.

**Comment by vpisarev** [comment]
:+1:

**Comment by K-Shinotsuka** [comment]
Thank you for reviewing and merging.

**Pull request merged**